flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -1794,9 +1816,12 @@ public void jobManagerLostLeadership(final JobID jobId, final JobMasterId jobMas
 			log.info("JobManager for job {} with leader id {} lost leadership.", jobId, jobMasterId);
 
 			runAsync(() ->
```

Comment:
Shouldn't we only run `disconnectJobManagerConnection` asynchronous? Just so we don't schedule something if no connection is present.

Reply:
In order to check whether a connection is present, we need to access the TM's state. This needs to happen from the main thread in order to be consistent.